File: test/language/functions/cpp/StructureArrayParamTest.cpp
#include <array>
#include <vector>

#include "functions/structure_array_param/ChildStructure.h"
#include "functions/structure_array_param/ParentStructure.h"
#include "gtest/gtest.h"
#include "zserio/BitStreamReader.h"
#include "zserio/BitStreamWriter.h"

namespace functions
{
namespace structure_array_param
{

class StructureArrayParamTest : public ::testing::Test
{
protected:
    void writeParentStructureToByteArray(zserio::BitStreamWriter& writer)
    {
        writer.writeBits(VALUES.size(), 8);

        for (uint64_t value : VALUES)
            writer.writeBits(static_cast<uint32_t>(value), CHILD_BIT_SIZE);

        writer.writeBits(ANOTHER_VALUES.size(), 8);

        for (uint64_t anotherValue : ANOTHER_VALUES)
            writer.writeBits(static_cast<uint32_t>(anotherValue), ANOTHER_CHILD_BIT_SIZE);
    }

    void createParentStructure(ParentStructure& parentStructure)
    {
        parentStructure.setNumChildren(VALUES.size());

        auto& children = parentStructure.getChildren();
        for (uint64_t value : VALUES)
        {
            ChildStructure child;
            child.setValue(value);
            children.push_back(child);
        }

        parentStructure.setNumAnotherChildren(ANOTHER_VALUES.size());

        auto& anotherChildren = parentStructure.getAnotherChildren();
        for (uint64_t anotherValue : ANOTHER_VALUES)
        {
            ChildStructure child;
            child.setValue(anotherValue);
            anotherChildren.push_back(child);
        }

        parentStructure.initializeChildren();
    }

    static const uint8_t CHILD_BIT_SIZE = 19;
    static const uint8_t ANOTHER_CHILD_BIT_SIZE = 17;

private:
    static const std::array<uint64_t, 2> VALUES;
    static const std::array<uint64_t, 2> ANOTHER_VALUES;
};

const std::array<uint64_t, 2> StructureArrayParamTest::VALUES = {0xAABB, 0xCCDD};

const std::array<uint64_t, 2> StructureArrayParamTest::ANOTHER_VALUES = {0xAABB, 0xCCDD};

TEST_F(StructureArrayParamTest, checkParentStructure)
{
    ParentStructure parentStructure;
    createParentStructure(parentStructure);
    const uint8_t expectedChildBitSize = CHILD_BIT_SIZE;
    ASSERT_EQ(expectedChildBitSize, parentStructure.funcGetChildBitSize());
    const uint8_t expectedAnotherChildBitSize = ANOTHER_CHILD_BIT_SIZE;
    ASSERT_EQ(expectedAnotherChildBitSize, parentStructure.getNotLeftMost().funcGetAnotherChildBitSize());

    zserio::BitBuffer writtenBitBuffer = zserio::BitBuffer(1024 * 8);
    zserio::BitStreamWriter writtenWriter(writtenBitBuffer);
    parentStructure.write(writtenWriter);

    zserio::BitBuffer expectedBitBuffer = zserio::BitBuffer(1024 * 8);
    zserio::BitStreamWriter expectedWriter(expectedBitBuffer);
    writeParentStructureToByteArray(expectedWriter);

    ASSERT_EQ(expectedBitBuffer, writtenBitBuffer);

    zserio::BitStreamReader reader(writtenBitBuffer);
    const ParentStructure readParentStructure(reader);
    ASSERT_EQ(parentStructure, readParentStructure);
}

} // namespace structure_array_param
} // namespace functions

File: test/language/functions/python/StructureArrayParamTest.py
import unittest
import zserio

from testutils import getZserioApi

class StructureArrayParamTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.api = getZserioApi(__file__, "functions.zs").structure_array_param

        cls.CHILD_BIT_SIZE = 19
        cls.ANOTHER_CHILD_BIT_SIZE = 17
        cls.CHILDREN = [cls.api.ChildStructure(cls.CHILD_BIT_SIZE, 0xAABB),
                        cls.api.ChildStructure(cls.CHILD_BIT_SIZE, 0xCCDD)]
        cls.ANOTHER_CHILDREN = [cls.api.ChildStructure(cls.ANOTHER_CHILD_BIT_SIZE, 0xAABB),
                        cls.api.ChildStructure(cls.ANOTHER_CHILD_BIT_SIZE, 0xCCDD)]
        cls.NUM_CHILDREN = len(cls.CHILDREN)
        cls.NUM_ANOTHER_CHILDREN = len(cls.ANOTHER_CHILDREN)

    def testParentStructure(self):
        parentStructure = self._createParentStructure()
        self.assertEqual(self.CHILD_BIT_SIZE, parentStructure.get_child_bit_size())
        self.assertEqual(self.ANOTHER_CHILD_BIT_SIZE,
                         parentStructure.not_left_most.get_another_child_bit_size())

        writer = zserio.BitStreamWriter()
        parentStructure.write(writer)
        expectedWriter = zserio.BitStreamWriter()
        self._writeParentStructureToStream(expectedWriter)
        self.assertTrue(expectedWriter.byte_array == writer.byte_array)
        self.assertTrue(expectedWriter.bitposition == writer.bitposition)

        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readParentStructure = self.api.ParentStructure.from_reader(reader)
        self.assertEqual(parentStructure, readParentStructure)

    def _writeParentStructureToStream(self, writer):
        writer.write_bits(self.NUM_CHILDREN, 8)
        for childStructure in self.CHILDREN:
            writer.write_bits(childStructure.value, childStructure.bit_size)

        writer.write_bits(self.NUM_ANOTHER_CHILDREN, 8)
        for childStructure in self.ANOTHER_CHILDREN:
            writer.write_bits(childStructure.value, childStructure.bit_size)

    def _createParentStructure(self):
        return self.api.ParentStructure(self.api.NotLeftMost(), self.NUM_CHILDREN, self.CHILDREN,
                                        self.NUM_ANOTHER_CHILDREN, self.ANOTHER_CHILDREN)
